### ISSUE TITLE
Cleaner RunningAverage syntax

### ISSUE THREAD

**Issue by miguelvr** [pytorch/ignite#278]
The current way to add multiple running averages computed over the output is a bit too verbose imo, and it can be made simpler.

## DCGAN Example:

```
    def step(engine, batch):

       (...)

        return {
            'errD': errD.item(),
            'errG': errG.item(),
            'D_x': D_x,
            'D_G_z1': D_G_z1,
            'D_G_z2': D_G_z2
        }
```

### Current API

To compute running averages over all of these, I would have to create an object per output and call the `attach` method:

```
    avg_err_d = RunningAverage(output_transform=lambda x: x['errD'])
    avg_err_d.attach(trainer, 'errD')
    avg_err_g = RunningAverage(output_transform=lambda x: x['errG'])
    avg_err_g.attach(trainer, 'errG')
    avg_dx = RunningAverage(output_transform=lambda x: x['D_x'])
    avg_dx.attach(trainer, 'D_x')
    avg_dg_z1 = RunningAverage(output_transform=lambda x: x['D_G_z1'])
    avg_dg_z1.attach(trainer, 'D_G_z1')
    avg_dg_z2 = RunningAverage(output_transform=lambda x: x['D_G_z2'])
    avg_dg_z2.attach(trainer, 'D_G_z2')
```

### Proposed API

Initialize one `RunningAverage` object and pass the `output_transform` function to the `attach` method

```
    running_avgs = RunningAverage()
    running_avgs.attach(trainer, 'errD', output_transform=lambda x: x['errD'])
    running_avgs.attach(trainer, 'errG', output_transform=lambda x: x['errG'])
    running_avgs.attach(trainer, 'D_x', output_transform=lambda x: x['D_x'])
    running_avgs.attach(trainer, 'D_G_z1', output_transform=lambda x: x['D_G_z1'])
    running_avgs.attach(trainer, 'D_G_z2', output_transform=lambda x: x['D_G_z2'])
```

cc @vfdev-5

**Comment by vfdev-5** [comment]
@miguelvr thanks, looks good! However as `RunningAverage` is a part of `Metric`, this makes me think about metric usage in general in the same way you propose. Today we have for example:
```
precision = Precision(average=True, output_transform=some_engine_specific_transform)
precision.attach(engine, "precision")
```
So for metrics it does not make too much sense to attach multiple times to the same engine. But it could make sense to attach to different engines:
```
precision.attach(trainer, "precision") # sort of precision on variable model
precision.attach(train_evaluator, "precision") # precision on the training dataset
precision.attach(val_evaluator, "precision") # precision on the validation dataset. Note that val_evaluator can have additional logic as early stopping if compared to train_evaluator.
```
however `output_transform` is defined once and is valable for a single engine. So maybe it makes sense to update `Metric` class as 
```
class Metric
    # Remove output_transform
    def __init__(self):
         pass
    def attach(self, engine, name, output_transform):
        pass
```

@jasonkriss any thoughts ?

**Comment by miguelvr** [comment]
On that issue, I found that there is no clear way of adding a single value as a metric, without accessing `engine.state.output` explicitly

**Comment by jasonkriss** [comment]
@miguelvr I see your point on the verbosity. I'm not sure how I feel about the proposal, though. In the `RunningAverage` use case, I think we lose some clarity and explicitness by allowing a single "metric" to be attached multiple times for various parts of the output. I personally find that confusing. The example could also be condensed like this:

```python
RunningAverage(output_transform=lambda x: x['errD']).attach(trainer, 'errD')
RunningAverage(output_transform=lambda x: x['errG']).attach(trainer, 'errG')
RunningAverage(output_transform=lambda x: x['D_x']).attach(trainer, 'D_x')
RunningAverage(output_transform=lambda x: x['D_G_z1']).attach(trainer, 'D_G_z1')
RunningAverage(output_transform=lambda x: x['D_G_z2']).attach(trainer, 'D_G_z2')
```

In which case, it's not really anymore verbose.

@vfdev-5 This makes a lil more sense to me. As in this case, precision is still a single "metric" it just happens to be computed on multiple datasets. In fact, it may even help prevent mistakes where a user would compute metrics differently across train and validation sets.

How do you envision this working with our current `create_supervised_*` functions?

I also wonder if maybe we should consider changing the return contract for `Engine.process_function` to be a dict instead of a tuple. E.g. in the supervised evaluator case, we could enforce that the the process function must return a dict with keys `y` and `y_pred`.  It would allow greater flexibility and perhaps obviate the issues brought up in #218. It might also lessen the need for custom output_transforms.

**Comment by vfdev-5** [comment]
> How do you envision this working with our current `create_supervised_*` functions?

Actually I do not see any incompatibility with `create_supervised_*` functions. Maybe I do not quite understand your question...

> I also wonder if maybe we should consider changing the return contract for Engine.process_function to be a dict instead of a tuple.

You mean for `create_supervised_*` functions? Seems reasonable and can be more explicit. Personnally, I got used to the list output.

**Comment by jasonkriss** [comment]
You're right. There is no compatibility. Not sure what i was thinking :).

Yea i mean in general, but since the `create_supervised_*` functions are the only explicit place where we define process functions, that's where it would show up.

**Comment by miguelvr** [comment]
i'm going to close this, as the condensed syntax proposed by @jasonkriss seems okay to me
